src/core/MIGRATION.md

```diff
@@ -604,9 +604,138 @@ It is generally a much greater challenge preparing legacy browser-side code for
 
-Also unlike the server-side migration, we won't concern ourselves with creating shimmed plugin definitions that then get copied over to complete the migration.
+Because usage of angular and `ui/public` modules varies widely between legacy plugins, there is no "one size fits all" solution to migrating your browser-side code to the new platform. The best place to start is by checking with the platform team to help identify the best migration path for your particular plugin.
 
-### Move UI modules into plugins
+That said, we've seen a series of patterns emerge as teams begin migrating browser code. In practice, most migrations will follow a path that looks something like this:
+
+#### 1. Create a plugin definition file
+
+We've found that doing this right away helps you start thinking about your plugin in terms of lifecycle methods and services, which makes the rest of the migration process feel more natural. It also forces you to identify which actions "kick off" your plugin, since you'll need to execute those when the `setup/start` methods are called.
+
+This definition isn't going to do much for us just yet, but as we get further into the process, we will gradually start returning contracts from our `setup` and `start` methods, while also injecting dependencies as arguments to these methods.
+
+```ts
+// public/plugin.ts
+import { CoreSetup, CoreStart, Plugin } from '../../../../core/public';
+import { FooSetup, FooStart } from '../../../../legacy/core_plugins/foo/public';
+
+/** @public */
+export type DemoSetup = {}
+/** @public */
+export type DemoStart = {}
+
+/** @internal */
+export interface DemoSetupPlugins {
+  foo: FooSetup;
+}
+/** @internal */
+export interface DemoStartPlugins {
+  foo: FooStart;
+}
+
+/** @internal */
+export class DemoPublicPlugin implements Plugin<DemoSetup, DemoStart, DemoSetupPlugins, DemoStartPlugins> {
+  public setup(core: CoreSetup, plugins: DemoSetupPlugins): DemoSetup {
+    // kick off your plugin here...
+    return {};
+  }
+
+  public start(core: CoreStart, plugins: DemoStartPlugins): DemoStart {
+    // ...or here
+    return {};
+  }
+
+  public stop() {}
+}
+```
+
+#### 2. Export all static code and types from `public/index.ts`
+
+If your plugin needs to share static code with other plugins, this code must be exported from your top-level `public/index.ts`. This includes any type interfaces that you wish to make public. For details on the types of code that you can safely share outside of the runtime lifecycle contracts, see [Can static code be shared between plugins?](#can-static-code-be-shared-between-plugins)
+
+```ts
+// public/index.ts
+import { DemoSetup, DemoStart } from './plugin';
```

Comment:
Would these be used by other plugins in their own `OtherDemoSetupPlugins` and `OtherDemoStartPlugins`?
I think its worth mentioning in a comment here.

Reply:
> Would these be used by other plugins

Yep exactly. I can add a comment